### PR TITLE
(MODULES-2706) types and providers

### PR DESCRIPTION
**Pull request by lucywyman**
Is there anything  else we need to add tests for, for types and providers?

I also wasn't sure whether to squash the commits or not -- let me know if I should!

### CODE REVIEW

File: spec/unit/type/logical_volume_spec.rb [hunk 1]
```diff
@@ -0,0 +1,337 @@
+#!/usr/bin/env rspec
+
+require 'spec_helper'
+describe Puppet::Type.type(:logical_volume) do
+	it 'raises an ArgumentError when the name has a file separator' do
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Looks like this file needs to switch to 2-space indentation

**Comment by lucywyman** [comment]
For documentation, total time: 3 hours

File: spec/unit/type/logical_volume_spec.rb [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+  end
+
+  it 'mirror number outside of 0-4 range throws error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name              => 'scooby doo',
+        :ensure            => :present,
+        :volume_group      => 'shaggy',
+        :size              => '10M',
+        :mirror            => -1
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Integers from puppet versions older than 4.0.0 can't actually pass integers; they only pass strings. Maybe check that strings work? (iirc anyway)

File: spec/unit/type/logical_volume_spec.rb [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        :ensure       => :present,
+        :volume_group => 'george',
+        :size         => '10M',
+        :minor        => 910
+      )
+    }.to raise_error(Puppet::ResourceError,
+                     'Parameter minor failed on Logical_volume[ringo]: 910 is not a valid value for minor. It must be an integer between 0 and 255')
+  end
+
+  it 'minor not set to integer between 0 and 255 raises error' do
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Description

File: spec/unit/type/logical_volume_spec.rb [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+  end
+
+  it 'range set within valid range does not raise error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name         => 'john',
+        :ensure        => :present,
+        :volume_group  => 'paul',
+        :size          => '10M',
+        # What is a valid range?
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: I think `range` only accepts `minimum` and `maximum`: https://www-01.ibm.com/support/knowledgecenter/ssw_aix_61/com.ibm.aix.osdevice/interdiskalloc.htm

File: spec/unit/type/logical_volume_spec.rb [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+  end
+
+  it 'invalid stripe count raises error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name         => 'bert',
+        :ensure        => :present,
+        :volume_group  => 'ernie',
+        :size          => '10M',
+        :stripes      => 12345
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: I believe this is actually a valid number of stripse (given 12345 physical volumes) so shouldn't throw an error.

File: spec/unit/type/logical_volume_spec.rb [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        :ensure        => :present,
+        :volume_group  => 'paul',
+        :size          => '10M',
+        # What is a valid range?
+        :range        => ['42', '0']
+      )
+    }.to_not raise_error
+  end
+
+  it 'invalid stripe count raises error' do
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: `s/stripe count/number of stripes/`

File: spec/unit/type/logical_volume_spec.rb [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        :ensure        => :present,
+        :volume_group  => 'ernie',
+        :size          => '10M',
+        :stripes      => 12345
+      )
+    }.to raise_error(Puppet::ResourceError,
+                     'Parameter stripes failed on Logical_volume[bert]: 12345 is not a valid stripe count')
+  end
+
+  it 'valid stripe count does not raise error' do
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Description

File: spec/unit/type/logical_volume_spec.rb [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name         => 'big bird',
+        :ensure        => :present,
+        :volume_group  => 'elmo',
+        :size          => '10M',
+        :stripes      => '7+'
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: 7+ is an invalid value and should raise an error. Only numbers or strings that are numbers are valid.

File: spec/unit/type/logical_volume_spec.rb [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+  end
+
+  it 'invalid stripesize raises error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name         => 'bert',
+        :ensure        => :present,
+        :volume_group  => 'ernie',
+        :size          => '10M',
+        :stripesize    => 12345
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Same as above

File: spec/unit/type/logical_volume_spec.rb [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+  end
+
+  it 'valid stripe count does not raise error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name         => 'big bird',
+        :ensure        => :present,
+        :volume_group  => 'elmo',
+        :size          => '10M',
+        :stripesize    => '7+'
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: same as above

File: spec/unit/type/logical_volume_spec.rb [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+  end
+
+  it 'region_size with invalid value throws error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name              => 'fred',
+        :ensure            => :present,
+        :volume_group      => 'daphne',
+        :size              => '10M',
+        :region_size      => '7+'
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Also should throw an error.

File: spec/unit/type/filesystem_spec.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        :options => '-b 4096 -E stride=32,stripe-width=64',
+      )
+    }.to raise_error(Puppet::ResourceError,
+                     'Parameter name failed on Filesystem[notAnAbsolutePath]: Filesystem names must be fully qualified')
+  end
+
+  it 'does not raise an ArgumentError when the name is an absolute path' do
+    expect {
+      resource = Puppet::Type.type(:filesystem).new(
+        :name => '/dev/mygv/mylv',
```
Comment (jordane) [inline spec/unit/type/filesystem_spec.rb]: `mygv`

Comment (lucywyman) [inline spec/unit/type/filesystem_spec.rb]: Hm?

File: spec/unit/type/logical_volume_spec.rb [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+  end
+
+  it 'valid stripesize does not raise error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name           => 'big bird',
+        :ensure         => :present,
+        :volume_group   => 'elmo',
+        :size           => '10M',
+        :stripesize     => '7'
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Could you update this to verify that `'7+'` DOES raise an error?

File: spec/unit/type/logical_volume_spec.rb [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+  end
+
+  it 'region_size set to positive integer does not throw error' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name             => 'fred',
+        :ensure           => :present,
+        :volume_group     => 'daphne',
+        :size             => '10M',
+        :region_size      => '910'
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: I think you had a `'7+'` value here that should throw an error. Could you do that again?

**Comment by hunner** [comment]
Looks like travis ci failure is unrelated to the changes in this pr. Maybe it's transient?

File: spec/unit/type/logical_volume_spec.rb (hunk 13 above)
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: And also another tests that passes a `7` to verify that integers are accepted?

File: spec/unit/type/filesystem_spec.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env rspec
+
+require 'spec_helper'
+
+describe Puppet::Type.type(:filesystem) do
+
+  it 'raises an ArgumentError when the name is not an absolute path' do
+    expect {
+      resource = Puppet::Type.type(:filesystem).new(
+        :name => 'notAnAbsolutePath',
```
Comment (bmjen) [inline spec/unit/type/filesystem_spec.rb]: rocket alignment.

File: spec/unit/type/filesystem_spec.rb [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        :options => '-b 4096 -E stride=32,stripe-width=64',
+      )
+    }.to raise_error(Puppet::ResourceError,
+                     'Parameter name failed on Filesystem[notAnAbsolutePath]: Filesystem names must be fully qualified')
+  end
+
+  it 'does not raise an ArgumentError when the name is an absolute path' do
+    expect {
+      resource = Puppet::Type.type(:filesystem).new(
+        :name => '/dev/myvg/mylv',
```
Comment (bmjen) [inline spec/unit/type/filesystem_spec.rb]: rockets

File: spec/unit/type/filesystem_spec.rb [hunk 17]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env ruby
```
Comment (lucywyman) [inline spec/unit/type/filesystem_spec.rb]: @hunner @bmjen  So this was the problematic line (it was #!/usr/bin/env rspec). Should we leave it as is, or revert to what's on master which is nothing?

**Comment by lucywyman** [comment]
https://travis-ci.org/puppetlabs/puppetlabs-lvm/jobs/90990611

These failures are expected, as we haven't addressed the '7+ doesn't fail' bug yet, and we (or Travis) is using a puppet version less than 4.0, so 7 instead of '7' should fail. 

Should I fix the '7+' bug here, or make a new issue and merge that first?

**Comment by hunner** [comment]
Yeah, the 7+ bug is something you found by making the tests better, so if you want to fix the validation in the type (it needs a `$` in the validation regex to match end of line) so the tests pass, that would be the preferred route.

File: spec/unit/type/logical_volume_spec.rb [hunk 18]
```diff
@@ -0,0 +1,373 @@
+require 'spec_helper'
+
+describe Puppet::Type.type(:logical_volume) do
+  it 'raises an ArgumentError when the name has a file separator' do
+    expect {
+      resource = Puppet::Type.type(:logical_volume).new(
+        :name         => '/dev/lol',
+        :ensure       => :present,
+        :volume_group => 'myvg',
+        :size         => '20G',
```
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: Just need to put {}'s around this hash to make ruby 1.8 pass

**Comment by lucywyman** [comment]
It's passing! It's passing!

**Comment by bmjen** [comment]
\o/

I think @hunner had a couple other suggestions.. and then a squash. Almost there!

File: spec/unit/type/logical_volume_spec.rb (hunk 13 above)
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: This is done.

File: spec/unit/type/logical_volume_spec.rb (hunk 14 above)
Comment (hunner) [inline spec/unit/type/logical_volume_spec.rb]: This is done.

**Pull request merged**